### PR TITLE
Fixes logging to scope is ignored for default listeners.

### PR DESCRIPTION
**Pull request by dereuromark**
Currently scope of `[]` defaults to `*`, which is derp. Found that out trying to script [executionorder demo app](https://github.com/dereuromark/executionorder). Can be reproduced there, as well.

When a scope is defined, the default listeners always now also always get to log, resulting in duplicate logs, no matter what.
This resolves it.

### CODE REVIEW

**Comment by ravage84** [comment]
Isn't this also happening in 2,x, too?
There we didn't fix it because it breaks "BC". :trollface: 
Hopefully we don't have to keep this illogical behavior for 3.x...

**Comment by dereuromark** [comment]
A major milestone is there to allow BC breaking.
I know that it is possible to have this behavior in 2.x, maybe some tiny little other way.
So I mainly see the issue here for 3.x right now.

**Comment by lorenzo** [comment]
I think this will need a special case for `scopes => ['*']` so that there is still the chance to listen to all scopes

**Comment by dereuromark** [comment]
If you want you can add a commit on top

**Comment by dereuromark** [comment]
The alternative would be to allow bool false maybe?

**Comment by markstory** [comment]
Why is a logger with no scopes, getting all messages 'derp'. Given that you have not defined a specific scope for the logger it should receive all 'messages'. This is how python's logging works as well.

**Comment by dereuromark** [comment]
Heh, I mainly cited @lorenzo there after a small chat in IRC. That word is usually not one that I am too familiar with ;)

The fact that they can receive all (by default) is not the problem, but the fact that this cannot be turned off at all.
When I want to just log to a single scope (and not want this to also end up in debug or error log) without losing the ability to still log everything else, this is currently not possible anymore.

**Comment by markstory** [comment]
The issue I have is that unscoped loggers are supposed to catch everything for a given level. Instead if adding implicit scopes have we considered adding a way for a logger to stop message delivery marking a message as completed?

**Comment by lorenzo** [comment]
@markstory I think the only feature we are missing is a way to toggle the "listen to all scopes" behavior. Currently we have a way to listen to all scopes, but to to prevent that.

**Comment by markstory** [comment]
Wouldn't defining specific scopes for a logger prevent it from getting all messages? Or is the issue that you want a logger to only get messages with no scope?

**Comment by dereuromark** [comment]
Neither. Unless I explicitly make a logger a catch-all one, I would like to not log scoped ones.
But I also like your idea:

> have we considered adding a way for a logger to stop message delivery marking a message as completed

But that would require that we can have a logger priority/order so that we can move some loggers up to the chain and those then to mark them. Otherwise default listeners would still first catch it before such a custom scoped one would be able to prevent it.

**Comment by dereuromark** [comment]
I updated with the ability to not catch all by default using `false`.
But why are the tests now failing? Shoudn't this be 100% BC?

**Comment by dereuromark** [comment]
Got it - $scopes can also be null, especially in the tests.

**Comment by markstory** [comment]
So with scope=false a logger will only receive unscoped messages? I think that is a good solution to the issue at hand.

File: src/Log/Log.php [hunk 1]
```diff
@@ -369,7 +369,9 @@ public static function write($level, $message, $context = [])
             }
 
             $correctLevel = empty($levels) || in_array($level, $levels);
-            $inScope = empty($scopes) || array_intersect($context['scope'], $scopes);
+            $inScope = $scopes === [] || $scopes === null ||
+                $scopes === false && empty($context['scope']) ||
```
Comment (markstory) [inline src/Log/Log.php]: Would it be simpler to check all the falsey things with in_array?

**Comment by dereuromark** [comment]
Exactly. :)

File: src/Log/Log.php (hunk 1 above)
Comment (dereuromark) [inline src/Log/Log.php]: The explicit $scopes = false needs a special check here.

Comment (lorenzo) [inline src/Log/Log.php]: this is difficult to read... I'd ask the same as markstory... ins't it the same as just saying `empty($scope)` ?

Comment (dereuromark) [inline src/Log/Log.php]: You could only simplify it to:
```php
if ($scopes === null) {
    $scopes = [];
}
$inScope = $scopes === false && empty($context['scope']) || array_intersect($context['scope'], $scopes);
```

Comment (markstory) [inline src/Log/Log.php]: That is easier to read imo.

**Comment by dereuromark** [comment]
Test failures are unrelated.

**Comment by AD7six** [comment]
@dereuromark you can fix the date errors with a rebase

**Comment by dereuromark** [comment]
Or we can simply continue in this case.

**Comment by markstory** [comment]
It would be nice to have a test for this behavior.

**Comment by dereuromark** [comment]
Test added.

**Comment by jadb** [comment]
:+1:

**Pull request merged**